scripts/setup.sh

```diff
@@ -0,0 +1,290 @@
+        echo "dnf"
+    elif command -v yum &>/dev/null; then
+        echo "yum"
+    elif command -v pacman &>/dev/null; then
+        echo "pacman"
+    else
+        echo "Error: Unsupported package manager" >&2
+        exit 1
+    fi
+}
+
+function install_package() {
+    local pkg_manager
+    pkg_manager=$(detect_package_manager)
+    
+    case $pkg_manager in
+        "apt")
+            apt-get update
+            apt-get install -y "$1"
+            ;;
+        "dnf")
+            dnf install -y "$1"
+            ;;
+        "yum")
+            yum install -y "$1"
+            ;;
+        "pacman")
+            pacman -Sy --noconfirm "$1"
+            ;;
+    esac
+}
+
+# check if the os is linux
+function check_os() {
+    if [ "$(uname)" != "Linux" ]; then
+        echo "Error: This script is only supported on Linux." >&2
+        exit 1
+    fi
+}
+
+# check if the script is running as root
+function check_root() {
+    if [ "$EUID" -ne 0 ]; then
+        echo "Error: Please run as root (sudo)" >&2
+        exit 1
+    fi
+}
+
+# check if the required commands are installed
+function check_command() {
+    local cmd="$1"
+    if ! command -v "$cmd" &>/dev/null; then
+        echo "Command '$cmd' not found. Attempting to install..."
+        case "$cmd" in
+            "git")
+                install_package "git"
+                ;;
+            "docker")
+                install_package "docker.io"
+                ;;
+            "yarn")
+                install_package "nodejs"
+                install_package "npm"
+                npm install -g yarn
+                ;;
+            *)
+                echo "Error: Automatic installation not supported for '$cmd'" >&2
+                exit 1
+                ;;
+        esac
+    fi
+}
+
+function check_required_commands() {
+    local commands=("git" "docker" "yarn")
+    for cmd in "${commands[@]}"; do
+        check_command "$cmd"
+    done
+}
+
+# clone the nixopus repository
+function clone_nixopus() {
+    if [ -d "nixopus" ]; then
+        echo "nixopus directory already exists. Removing..."
+        rm -rf nixopus
+    fi
+    if ! git clone https://github.com/raghavyuva/nixopus.git; then
+        echo "Error: Failed to clone nixopus repository" >&2
+        exit 1
+    fi
+}
+
+# checkout to the branch
+function checkout_branch() {
+    local branch="$1"
+    if ! git checkout "$branch"; then
+        echo "Error: Failed to checkout to $branch" >&2
+        exit 1
+    fi
+}
+
+# move to the folder
+function move_to_folder() {
+    local folder="$1"
+    if ! cd "$folder"; then
+        echo "Error: Failed to change directory to $folder" >&2
+        exit 1
+    fi
+}
+
+function install_go() {
+    local version="1.23.4"
+    local arch
+    arch=$(uname -m)
+    case "$arch" in
+        x86_64) arch="amd64" ;;
+        aarch64|arm64) arch="arm64" ;;
+        *) echo "Unsupported architecture: $arch" >&2; exit 1 ;;
+    esac
+    local os="linux"
```

Comment:
We can support other OSes right? Like macos can directly use brew

Reply:
Yes, this is just a starting point which supports linux for now, we can extend this once we have working Linux systems setup